crates/models/gptneox/src/lib.rs

```diff
@@ -182,95 +216,55 @@ impl KnownModel for GptNeoX {
-                    &qcur,
-                    &ctx0.new_tensor_3d(ElementType::F32, n_embd / n_head, n_head, n),
-                );
-                kcur = ctx0.op_cpy(
-                    &kcur,
-                    &ctx0.new_tensor_3d(ElementType::F32, n_embd / n_head, n_head, n),
-                );
-                vcur = ctx0.op_cpy(
-                    &vcur,
-                    &ctx0.new_tensor_3d(ElementType::F32, n_embd / n_head, n_head, n),
-                );
-            }
+            let nb = current.get_nb()[1];
+            let f32_size = std::mem::size_of::<f32>();
+
+            let mut qcur = ctx0.op_cont(&ctx0.op_view_3d(
+                &current,
+                (n_embd / n_head, n_head, n),
+                (nb / n_head, nb),
+                0,
+            ));
+            let mut kcur = ctx0.op_cont(&ctx0.op_view_3d(
+                &current,
+                (n_embd / n_head, n_head, n),
+                (nb / n_head, nb),
+                f32_size * n_embd / n_head,
+            ));
+            let mut vcur = ctx0.op_cont(&ctx0.op_view_3d(
+                &current,
+                (n_embd / n_head, n_head, n),
+                (nb / n_head, nb),
+                2 * f32_size * n_embd / n_head,
+            ));
 
             // self-attention using mode = 2 for GPT-NeoX mode
             qcur = ctx0.op_rope_inplace(&qcur, n_past, n_rot, 2);
             kcur = ctx0.op_rope_inplace(&kcur, n_past, n_rot, 2);
 
-            // self-attention store key and value to memory
-            if use_parallel_residual {
-                vcur = ctx0.op_transpose(&ctx0.op_reshape_2d(&vcur, n_embd, n));
-            } else {
-                vcur = ctx0.op_view_2d(&vcur, (n_embd, n), vcur.element_size() * n_embd, 0);
-                vcur = ctx0.op_transpose(&vcur);
-            }
+            // store key and value to memory
+            vcur = ctx0.op_transpose(&ctx0.op_reshape_2d(&vcur, n_embd, n));
 
-            let little_k = ctx0.op_view_1d(
+            let k = ctx0.op_view_1d(
                 memory_k,
                 n * n_embd,
                 (memory_k_size * n_embd) * (il * n_ctx + n_past),
             );
-            let little_v = ctx0.op_view_2d(
+
+            let v = ctx0.op_view_2d(
                 memory_v,
                 (n, n_embd),
                 n_ctx * memory_v_size,
                 (il * n_ctx) * memory_v_size * n_embd + n_past * memory_v_size,
             );
 
-            gf.build_forward_expand(&ctx0.op_cpy(&kcur, &little_k));
-            gf.build_forward_expand(&ctx0.op_cpy(&vcur, &little_v));
+            gf.build_forward_expand(&ctx0.op_cpy(&kcur, &k));
+            gf.build_forward_expand(&ctx0.op_cpy(&vcur, &v));
 
-            let q = ctx0.op_permute(&qcur, 0, 2, 1, 3);
-            let big_k = ctx0.op_permute(
+            // Q = Qcur.contiguous().view(n_embd/n_head, n_head, N).permute(0, 2, 1, 3)
+            let Q = ctx0.op_permute(&qcur, 0, 2, 1, 3);
```

Comment:
I'd suggest still lowercasing the names, if only because Rust syntax highlighting assumes that anything uppercase is a constant

Reply:
Well i actually don't know about this, at some point in the future we should/have to replace this with some attention head abstractions until then keeping it in sync with upstream is easier, if we keep the same names.

Reply:
Agreed, but will downcasing cause any issues? It should be pretty obvious they're the same thing... but if there's at all any conflicts then yeah leave it as-is, I'm not that picky about this